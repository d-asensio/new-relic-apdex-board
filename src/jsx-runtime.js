export default function createElement (tag, props, ...children) {
  const element = document.createElement(tag)

  if (props !== null) {
    _addPropsToElement(props, element)
  }

  _appendChildrenToElement(children, element)

  return element
}

function _addPropsToElement (props, element) {
  const { className, ...attributes } = props

  _addAttributesToElement(attributes, element)

  if (className) {
    _addClassNameToElement(className, element)
  }
}

function _addAttributesToElement (attributes, element) {
  const attributeEntries = Object.entries(attributes)

  for (const [attrName, attrValue] of attributeEntries) {
    _attributeSupportedByElementOrThrow(attrName, element)

    element.setAttribute(attrName, attrValue)
  }
}

function _attributeSupportedByElementOrThrow (attributeName, element) {
  if (!_isAttributeSupportedByElement(attributeName, element)) {
    throw new Error(
      `A "${element.tagName.toLowerCase()}" element do not support the attribute "${attributeName}".`
    )
  }
}

function _isAttributeSupportedByElement (attributeName, element) {
  if (_isDataAttribute(attributeName)) return true
  if (_isAriaAttribute(attributeName)) return true

  return attributeName in element
}

function _isDataAttribute (attributeName) {
  return attributeName.startsWith('data-')
}

function _isAriaAttribute (attributeName) {
  return attributeName.startsWith('aria-')
}

function _addClassNameToElement (className, element) {
  const sanitizedClassNames = _sanitizeClassName(className)

  for (const sanitizedClass of sanitizedClassNames) {
    element.classList.add(sanitizedClass)
  }
}

function _sanitizeClassName (className) {
  return className
    .split(' ')
    .filter(Boolean)
}

function _appendChildrenToElement (children, element) {
  for (const child of children) {
    if (Array.isArray(child)) {
      _appendChildrenToElement(child, element)
    } else {
      _appendChildToElement(child, element)
    }
  }
}

function _appendChildToElement (child, element) {
  const safeChildNode = _createSafeChildNode(child, element)

  element.appendChild(safeChildNode)
}

function _createSafeChildNode (child) {
  if (child instanceof HTMLElement || child instanceof Text) {
    return child
  }

  return document.createTextNode(child)
}
