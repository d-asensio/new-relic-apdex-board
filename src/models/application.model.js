import { uniqueId } from '../utils'

export class Application {
  constructor ({ name, apdex, version, contributors }) {
    this._id = uniqueId('app_')

    this._name = name
    this._apdex = apdex
    this._version = version
    this._contributors = contributors
  }

  get id () { return this._id }
  get name () { return this._name }
  get apdex () { return this._apdex }
  get version () { return this._version }
  get contributors () { return this._contributors }

  compareTo (comparee) {
    return this.apdex - comparee.apdex
  }

  static byApdexInDescOrder (appA, appB) {
    return appB.apdex - appA.apdex
  }
}
