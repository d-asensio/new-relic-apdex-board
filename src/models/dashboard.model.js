export class Dashboard {
  constructor ({ user }) {
    this.user = user

    this._hosts = new Map()
    this._apps = new Map()
  }

  addApp (app) {
    this._apps.set(app.id, app)
  }

  hasApp (appId) {
    return this._apps.has(appId)
  }

  getApp (appId) {
    return this._apps.get(appId)
  }

  addHost (host) {
    this._hosts.set(host.id, host)
  }

  hasHost (hostId) {
    return this._hosts.has(hostId)
  }

  getHost (hostId) {
    return this._hosts.get(hostId)
  }

  getAllHosts () {
    return Array.from(this._hosts.values())
  }
}
