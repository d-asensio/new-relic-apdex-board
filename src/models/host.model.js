export class Host {
  constructor ({ id }) {
    this.id = id
    this._apps = []
  }

  addApp (app) {
    this._insertAppInPosition(app)
  }

  removeApp (app) {
    this._removeAppFromPosition(app)
  }

  getTopApps (nTopApps) {
    return this._apps.slice(0, nTopApps)
  }

  _insertAppInPosition (app) {
    const appInsertionIndex = this._binarySearch(app)
    this._insertAppAtIndex(app, appInsertionIndex)
  }

  _removeAppFromPosition (app) {
    const appIndex = this._binarySearch(app)

    this._indexExistOrThrow(appIndex)
    this._removeAppAtIndex(appIndex)
  }

  _binarySearch (app) {
    let startIndex = 0
    let endIndex = this._apps.length - 1
    let currentIndex

    while (startIndex <= endIndex) {
      currentIndex = (startIndex + endIndex) / 2 | 0
      const iteratingApp = this._apps[currentIndex]

      if (app.compareTo(iteratingApp) < 0) {
        startIndex = currentIndex + 1
      } else if (app.compareTo(iteratingApp) > 0) {
        endIndex = currentIndex - 1
      } else {
        return currentIndex
      }
    }

    return Math.abs(~endIndex)
  }

  _insertAppAtIndex (app, insertionIndex) {
    this._apps.splice(insertionIndex, 0, app)
  }

  _removeAppAtIndex (deletionIndex) {
    this._apps.splice(deletionIndex, 1)
  }

  _indexExistOrThrow (index) {
    if (!this._indexExists(index)) {
      throw new Error(
        'The specified app do not exist in the host.'
      )
    }
  }

  _indexExists (index) {
    return index >= 0 && index < this._apps.length
  }
}
