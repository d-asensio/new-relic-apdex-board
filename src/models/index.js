export * from './application.model.js'
export * from './dashboard.model.js'
export * from './host.model.js'
