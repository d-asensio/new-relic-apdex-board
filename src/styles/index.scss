@import "reset";

@import "tokens/all";
@import "utilities/all";

@import "views/all";

body {
  background: #f5f5f5;
  
  @include sp-inset($space--large);
}
