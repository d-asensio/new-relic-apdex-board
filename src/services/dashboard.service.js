import { Dashboard, Application, Host } from '../models'

const API_URL = 'data/host-app-data.json'

/**
 * Manages the data of the dashboard, meaning that it kwows how to create models and how to get data to create them.
 */
export class DashboardService {
  constructor () {
    this._dashboard = new Dashboard({
      user: 'this.email.is.so.long@newrelic.com'
    })
  }

  get dashboard () { return this._dashboard }

  async init () {
    const appRecords = await this._fetchAppRecordsFromApi()

    const sortedRecords = this._sortAppRecordsByApdex(appRecords)
    this._createAppModelsFromRecords(sortedRecords)
  }

  addAppToHosts (app, hosts) {
    this._appExistsOrThrow(app)

    for (const hostId of hosts) {
      const host = this._getOrCreateHostById(hostId)
      host.addApp(app)
    }
  }

  removeAppFromHosts (app, hosts) {
    this._appExistsOrThrow(app)

    for (const hostId of hosts) {
      this._hostIdExistsOrThrow(hostId)

      const host = this._getHostById(hostId)
      host.removeApp(app)
    }
  }

  getTopAppsByHost (hostId) {
    this._hostIdExistsOrThrow(hostId)

    const host = this._getHostById(hostId)

    return host.getTopApps(25)
  }

  getAppById (appId) {
    return this._dashboard.getAppById(appId)
  }

  async _fetchAppRecordsFromApi () {
    const res = await fetch(API_URL)
    return res.json()
  }

  _sortAppRecordsByApdex (appRecords) {
    return appRecords.sort(Application.byApdexInDescOrder)
  }

  _createAppModelsFromRecords (appRecords) {
    for (const { host, ...appData } of appRecords) {
      const app = new Application(appData)

      this._dashboard.addApp(app)
      this.addAppToHosts(app, host)
    }
  }

  _getOrCreateHostById (hostId) {
    if (!this._hostIdExists(hostId)) {
      return this._createHostById(hostId)
    }

    const host = this._getHostById(hostId)

    return host
  }

  _getHostById (hostId) {
    return this._dashboard.getHostById(hostId)
  }

  _createHostById (hostId) {
    const host = new Host({ id: hostId })

    this._dashboard.addHost(host)

    return host
  }

  _hostIdExistsOrThrow (hostId) {
    if (!this._hostIdExists(hostId)) {
      throw new Error(
        `The host identified by "${hostId} do not exist."`
      )
    }
  }

  _appExistsOrThrow (app) {
    if (!this._appExists(app)) {
      throw new Error(
        `The app ${app.name} do not exist in the dashboard.`
      )
    }
  }

  _hostIdExists (hostId) {
    return this._dashboard.hasHostId(hostId)
  }

  _appExists (app) {
    return this._dashboard.hasAppId(app.id)
  }
}
