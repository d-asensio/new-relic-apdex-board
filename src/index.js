
import './styles/index.scss'

import { DashboardController } from './controllers'
import { DashboardService } from './services'
import { DashboardView } from './views'

const controller = new DashboardController(
  new DashboardService(),
  new DashboardView()
)

function renderApp () {
  const appRoot = document.getElementById('app')
  controller.renderToDom(appRoot)
}

;(async function IIFE () {
  await controller.init()

  if (document.readyState === 'complete') {
    renderApp()
  } else {
    document.addEventListener(
      'DOMContentLoaded',
      () => renderApp()
    )
  }
})()
