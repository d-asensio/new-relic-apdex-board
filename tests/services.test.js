import { enableFetchMocks } from 'jest-fetch-mock'
import mockData from '../data/host-app-data.json'

import { DashboardService } from '../src/services'
import { Dashboard, Application } from '../src/models'

enableFetchMocks()

const largeResponseMock = JSON.stringify(mockData.slice(0, 100))
const smallResponseMock = JSON.stringify(mockData.slice(0, 10))

describe('Dashboard service', () => {
  beforeEach(() => {
    fetch.resetMocks()
  })

  it('is constructed', () => {
    const dashboardService = new DashboardService()

    expect(dashboardService).toBeInstanceOf(DashboardService)
  })

  it('fetches data from the API when it is initialized', async () => {
    fetch.mockResponseOnce(largeResponseMock)

    const dashboardService = new DashboardService()
    await dashboardService.init()

    expect(fetch.mock.calls.length).toEqual(1)
  })

  it('throws an error when trying to get the top apps of a unexisting host', () => {
    const dashboardService = new DashboardService()

    expect(() => {
      dashboardService.getTopAppsByHost('unexisting.host.com')
    }).toThrowErrorMatchingSnapshot()
  })

  it('returns all the apps if there are less than 25 for a specified host', async () => {
    fetch.mockResponseOnce(smallResponseMock)

    const dashboardService = new DashboardService()
    await dashboardService.init()

    const topApps = dashboardService.getTopAppsByHost('7e6272f7-098e.dakota.biz')

    expect(topApps.length).toEqual(2)
  })

  it('returns only 25 apps for a specified host', async () => {
    fetch.mockResponseOnce(largeResponseMock)

    const dashboardService = new DashboardService()
    await dashboardService.init()

    const topApps = dashboardService.getTopAppsByHost('7e6272f7-098e.dakota.biz')

    expect(topApps.length).toEqual(25)
  })

  it('returns an app by id', async () => {
    fetch.mockResponseOnce(smallResponseMock)

    const dashboardService = new DashboardService()
    await dashboardService.init()

    const [appA] = dashboardService.getTopAppsByHost('7e6272f7-098e.dakota.biz')
    const gotApp = dashboardService.getAppById(appA.id)

    expect(gotApp).toBe(appA)
  })

  it('has dashboard instance', () => {
    const dashboardService = new DashboardService()

    expect(dashboardService.dashboard).toBeInstanceOf(Dashboard)
  })

  it('the dashboard instance is not modificable', () => {
    const dashboardService = new DashboardService()

    expect(() => {
      dashboardService.dashboard = new Dashboard({
        user: 'some.user@newrelic.com'
      })
    }).toThrowError()
  })

  it('adds an app to a new host', async () => {
    fetch.mockResponseOnce(JSON.stringify([
      {
        name: 'A',
        contributors: [],
        version: 7,
        apdex: 68,
        host: [
          'first.host.com'
        ]
      },
      {
        name: 'B',
        contributors: [],
        version: 7,
        apdex: 30,
        host: [
          'first.host.com'
        ]
      }
    ]))

    const dashboardService = new DashboardService()
    await dashboardService.init()

    const [appA] = dashboardService.getTopAppsByHost('first.host.com')

    dashboardService.addAppToHosts(appA, [
      'second.host.com'
    ])

    const [appASecond] = dashboardService.getTopAppsByHost('second.host.com')

    expect(appASecond).toBe(appA)
  })

  it('removes an app from a host', async () => {
    fetch.mockResponseOnce(JSON.stringify([
      {
        name: 'A',
        contributors: [],
        version: 7,
        apdex: 68,
        host: [
          'first.host.com'
        ]
      },
      {
        name: 'B',
        contributors: [],
        version: 7,
        apdex: 30,
        host: [
          'first.host.com'
        ]
      }
    ]))

    const dashboardService = new DashboardService()
    await dashboardService.init()

    const [appA, appB] = dashboardService.getTopAppsByHost('first.host.com')

    dashboardService.removeAppFromHosts(appA, [
      'first.host.com'
    ])

    const [appBFirst] = dashboardService.getTopAppsByHost('first.host.com')

    expect(appBFirst).toBe(appB)
  })

  it('attempts to add an app that do not exist', async () => {
    fetch.mockResponseOnce('[]')

    const dashboardService = new DashboardService()
    await dashboardService.init()

    const unexistingApp = new Application({
      name: 'A',
      apdex: 33,
      version: 7,
      contributors: []
    })

    expect(() => {
      dashboardService.addAppFromHosts(unexistingApp, [
        'first.host.com'
      ])
    }).toThrowErrorMatchingSnapshot()
  })

  it('attempts to remove an app that do not exist', async () => {
    fetch.mockResponseOnce('[]')

    const dashboardService = new DashboardService()
    await dashboardService.init()

    const unexistingApp = new Application({
      name: 'A',
      apdex: 33,
      version: 7,
      contributors: []
    })

    expect(() => {
      dashboardService.removeAppFromHosts(unexistingApp, [
        'first.host.com'
      ])
    }).toThrowErrorMatchingSnapshot()
  })

  it('continues returning 25 top apps after removing', async () => {
    fetch.mockResponseOnce(largeResponseMock)

    const dashboardService = new DashboardService()
    await dashboardService.init()

    const [highestApexApp] = dashboardService.getTopAppsByHost('7e6272f7-098e.dakota.biz')

    dashboardService.removeAppFromHosts(highestApexApp, [
      '7e6272f7-098e.dakota.biz'
    ])

    const topApps = dashboardService.getTopAppsByHost('7e6272f7-098e.dakota.biz')

    expect(topApps.length).toEqual(25)
  })
})
