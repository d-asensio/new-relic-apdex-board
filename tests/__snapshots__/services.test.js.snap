// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard service attempts to add an app that do not exist 1`] = `"dashboardService.addAppFromHosts is not a function"`;

exports[`Dashboard service attempts to remove an app that do not exist 1`] = `"The app A do not exist in the dashboard."`;

exports[`Dashboard service throws an error when trying to get the top apps of a unexisting host 1`] = `"The host identified by \\"unexisting.host.com do not exist.\\""`;
